compiler/rustc_query_system/src/query/job.rs

```diff
@@ -28,76 +30,78 @@ use {
 
 /// Represents a span and a query key.
 #[derive(Clone, Debug)]
-pub struct QueryInfo {
+pub struct QueryInfo<D: DepKind> {
     /// The span corresponding to the reason for which this query was required.
     pub span: Span,
-    pub query: QueryStackFrame,
+    pub query: QueryStackFrame<D>,
 }
 
-pub type QueryMap = FxHashMap<QueryJobId, QueryJobInfo>;
+pub type QueryMap<D> = FxHashMap<QueryJobId<D>, QueryJobInfo<D>>;
 
 /// A value uniquely identifying an active query job.
 #[derive(Copy, Clone, Eq, PartialEq, Hash)]
-pub struct QueryJobId(pub NonZeroU64);
+pub struct QueryJobId<D: DepKind>(pub NonZeroU64, pub core::marker::PhantomData<D>);
```

Comment:
This doesn't need to be generic, right? You can just make the methods generic instead which reduces the amount of generics that have to be carried around.